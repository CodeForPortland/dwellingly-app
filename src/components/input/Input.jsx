import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import RadioArray from './RadioArray';
import Toggle from './Toggle';

import './Input.scss';

function InputContent({
  blockClass,
  children,
  id, // identifier for use with htmlFor and key properties
  label,
  maxLength, // maximum length
  model, // value of key in state
  name, // name of input
  onChange, // passes along the onChange function through props
  onClick, // for use with button/Link
  placeholder,
  type, // defines input type ["radio", "button", etc.]
  url, // for use with button/Link
  value // value from state, object, or string
}) {
  let inputContent = (
    <label htmlFor={id}>
      <span className={`${blockClass}__label`}>{label}</span>
      <input
        id={id}
        maxLength={maxLength || 200}
        name={name}
        onChange={onChange}
        placeholder={placeholder}
        type={type}
        value={value}
      />
    </label>
  );

  if (type === 'radio') {
    inputContent = (
      <label htmlFor={id}>
        <input
          checked={value === model}
          id={id}
          model={model}
          name={name}
          onChange={onChange}
          type="radio"
          value={value}
        />
        <span className="inline-input__radio-btn" />
        {label}
      </label>
    );
  }

  if (type === 'button') {
    inputContent = (
      <div>
        {!url && (
          <button aria-label={label} type="button" onClick={onClick}>
            {children}
          </button>
        )}
        {url && <Link to={url}>{children}</Link>}
      </div>
    );
  }

  if (type === 'date') {
    inputContent = (
      <label htmlFor={id}>
        <span className={`${blockClass}__label`}>{label}</span>
        <input
          id={id}
          name={name}
          onChange={onChange}
          type="datetime-local"
          value={value}
        />
      </label>
    );
  }

  return inputContent;
}

export default class Input extends Component {
  render() {
    const { blockClass, type, variants } = this.props;

    const variantsClasses =
      variants.length > 0
        ? variants
            .map(variant => `${blockClass}__${type}--${variant}`)
            .join(' ')
        : '';

    return (
      <div className={`input ${blockClass}`}>
        <div className={`${blockClass}__${type} ${variantsClasses}`}>
          <InputContent {...this.props} variantsClasses={variantsClasses} />
        </div>
      </div>
    );
  }
}

Input.propTypes = {
  blockClass: PropTypes.string,
  type: PropTypes.string,
  variants: PropTypes.arrayOf(PropTypes.string)
};

Input.defaultProps = {
  blockClass: 'inline-input',
  type: 'text',
  variants: []
};

export { Input, RadioArray, Toggle };
