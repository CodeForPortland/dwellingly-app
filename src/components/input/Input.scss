@import "../../constants/constants.scss";

/* F O R M S */
/* ==================================================== */
// INLINE-INPUT
.inline-input {
  // adding variable here as we use it for nearly all variants of .inline-input
  $inline-input__padding: 0.75rem 0;

  @include flex_align(space-between);
  border-bottom: 1px solid $color__gray--border;
  font-size: 1rem;
  margin: 0 1rem;
  position: relative;
  text-align: left;

  &:last-of-type {
    border: 0;
  }

  /* BUTTON */
  &__btn {
    text-align: center;
    width: 100%;

    a,
    button {
      display: block;
      font-size: 1em;
      padding: $inline-input__padding;
      text-align: center;
      width: 100%;
    }
  }

  /* RADIO BUTTON */
  &__radio {
    width: 100%;
    label {
      cursor: pointer;
      display: block;
      padding: $inline-input__padding;
      padding-left: 3em;
      position: relative;
    }

    input[type="radio"] {
      position: absolute;
      left: 0;
      opacity: 0;

      &:checked + .inline-input__radio-btn {
        background: $color__yellow;
      }
    }

    &-btn {
      border: 2px solid $color__yellow;
      border-radius: 50%;
      display: block;
      height: 1.25em;
      left: 0.75em;
      position: absolute;
      top: 0.825em;
      transition: background 0.2s ease;
      width: 1.25em;
    }
  }

  /* TEXT */
  &__tel,
  &__text,
  &__date,
  &__email,
  &__password {
    padding: 0.25em 0;
    width: 100%;

    &--full {
      label {
        @include flex_align(space-between);
      }

      .inline-input__label {
        display: inline-block;
        padding-right: 1em;
        width: 30%;
      }
      
      input {
        width: 70%;
      }
    }
  }

  /* DATE */
  &__date {
    padding: $inline-input__padding;

    input {
      font-weight: inherit;
    }
  }

  /* "TOGGLE" CHECKBOX */
  &__toggle {
    padding: 0.5rem 0;
    input[type="checkbox"],
    input[type="radio"] {
      display: none;

      &:checked + .inline-input__toggle__btn {
        background: $color__green;
        border-color: $color__green;

        &::after {
          left: 1.5em;
        }

        &:active {
          &::after {
            margin-left: -0.5em;
          }
        }
      }
    }

    &__btn {
      background: $color__white;
      border: 1px solid rgba(0, 0, 0, 0.1);
      border-radius: 2em;
      display: block;
      font-size: 1em;
      height: 2em;
      position: relative;
      text-indent: -9999em;
      transition: background 0.2s ease, border 0.2s ease;
      width: 3.5em;

      &::after {
        background: $color__white;
        border: 1px solid rgba(0, 0, 0, 0.1);
        border-radius: 1.5em;
        box-shadow: 0 3px 3px rgba(0, 0, 0, 0.05), 0 2px 2px rgba(0, 0, 0, 0.1),
          0 3px 1px rgba(0, 0, 0, 0.05);
        content: "";
        display: block;
        height: calc(2em - 2px);
        left: -1px;
        padding: 0;
        position: absolute;
        transition: all 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
        top: -1px;
        width: calc(2em - 2px);
        z-index: 1;
      }

      &:hover::after {
        will-change: padding;
      }

      &:active {
        &::after {
          padding-right: 0.5em;
        }
      }
    }
  }
}
/* end forms */
