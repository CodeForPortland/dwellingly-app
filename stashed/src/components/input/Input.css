/* V A R I A B L E S */
/* end variables */
/* F U N C T I O N S */
/* end functions */
/* M I X I N S */
/* end mixins */
/* F O R M S */
/* ==================================================== */
.inline-input {
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
      -ms-flex-pack: justify;
          justify-content: space-between;
  border-bottom: 1px solid #dddddd;
  font-size: 1rem;
  margin: 0 1rem;
  position: relative;
  text-align: left;
  /* BUTTON */
  /* RADIO BUTTON */
  /* TEXT */
  /* DATE */
  /* "TOGGLE" CHECKBOX */
}

.inline-input:last-of-type {
  border: 0;
}

.inline-input.input--no-label:last-of-type {
  border-bottom: 1px solid #dddddd;
}

.inline-input.input--no-label .inline-input__label {
  display: none;
}

.inline-input.input--no-label input {
  width: 100%;
}

.inline-input__button {
  text-align: center;
  width: 100%;
}

.inline-input__button a,
.inline-input__button button {
  display: block;
  font-size: 1em;
  padding: 0.75rem 0;
  text-align: center;
  width: 100%;
}

.inline-input__radio {
  width: 100%;
}

.inline-input__radio label {
  cursor: pointer;
  display: block;
  padding: 0.75rem 0;
  padding-left: 3em;
  position: relative;
}

.inline-input__radio input[type='radio'] {
  position: absolute;
  left: 0;
  opacity: 0;
}

.inline-input__radio input[type='radio']:checked + .inline-input__radio-btn {
  background: #ffbd3b;
}

.inline-input__radio-btn {
  border: 2px solid #ffbd3b;
  border-radius: 50%;
  display: block;
  height: 1.25em;
  left: 0.75em;
  position: absolute;
  top: 0.825em;
  -webkit-transition: background 0.2s ease;
  transition: background 0.2s ease;
  width: 1.25em;
}

.inline-input__tel, .inline-input__text, .inline-input__date, .inline-input__email, .inline-input__number, .inline-input__password {
  padding: 0.25em 0;
  width: 100%;
}

.inline-input__tel label, .inline-input__text label, .inline-input__date label, .inline-input__email label, .inline-input__number label, .inline-input__password label {
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
      -ms-flex-pack: justify;
          justify-content: space-between;
}

.inline-input__tel .inline-input__label, .inline-input__text .inline-input__label, .inline-input__date .inline-input__label, .inline-input__email .inline-input__label, .inline-input__number .inline-input__label, .inline-input__password .inline-input__label {
  display: inline-block;
  padding-right: 1em;
  width: 30%;
}

.inline-input__tel input, .inline-input__text input, .inline-input__date input, .inline-input__email input, .inline-input__number input, .inline-input__password input {
  width: 70%;
}

.inline-input__date {
  padding: 0.75rem 0;
}

.inline-input__date input {
  font-weight: inherit;
  width: auto;
}

.inline-input__toggle {
  -webkit-box-align: center;
      -ms-flex-align: center;
          align-items: center;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
      -ms-flex-pack: justify;
          justify-content: space-between;
  padding: 0.5rem 0;
  width: 100%;
}

.switch input[type='checkbox'] {
  display: none;
}

.switch input[type='checkbox']:checked + .switch__btn {
  background: #4cd964;
  border-color: #4cd964;
}

.switch input[type='checkbox']:checked + .switch__btn::after {
  left: 1.5em;
}

.switch input[type='checkbox']:checked + .switch__btn:active::after {
  margin-left: -0.5em;
}

.switch__btn {
  background: #ffffff;
  border: 1px solid rgba(0, 0, 0, 0.1);
  border-radius: 2em;
  display: block;
  font-size: 1em;
  height: 2em;
  position: relative;
  text-indent: -9999em;
  -webkit-transition: background 0.2s ease, border 0.2s ease;
  transition: background 0.2s ease, border 0.2s ease;
  width: 3.5em;
}

.switch__btn::after {
  background: #ffffff;
  border: 1px solid rgba(0, 0, 0, 0.1);
  border-radius: 1.5em;
  -webkit-box-shadow: 0 3px 3px rgba(0, 0, 0, 0.05), 0 2px 2px rgba(0, 0, 0, 0.1), 0 3px 1px rgba(0, 0, 0, 0.05);
          box-shadow: 0 3px 3px rgba(0, 0, 0, 0.05), 0 2px 2px rgba(0, 0, 0, 0.1), 0 3px 1px rgba(0, 0, 0, 0.05);
  content: '';
  display: block;
  height: calc(2em - 2px);
  left: -1px;
  padding: 0;
  position: absolute;
  -webkit-transition: all 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
  transition: all 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
  top: -1px;
  width: calc(2em - 2px);
  z-index: 1;
}

.switch__btn:hover::after {
  will-change: padding;
}

.switch__btn:active::after {
  padding-right: 0.5em;
}

/* end forms */
/*# sourceMappingURL=Input.css.map */