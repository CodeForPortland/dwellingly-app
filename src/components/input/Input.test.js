import React from 'react';
import renderer from 'react-test-renderer';
import Input from './Input';

describe('Input Component', () => {
  it('correctly renders Input component', () => {
    const InputComponent = renderer.create(<Input />).toJSON();
    expect(InputComponent).toMatchSnapshot();
  });
});
