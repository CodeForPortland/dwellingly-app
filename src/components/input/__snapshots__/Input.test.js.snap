// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component correctly renders Input component 1`] = `
<div
  className="input inline-input"
>
  <div
    className="inline-input__text "
  >
    <label>
      <span
        className="inline-input__label"
      />
      <input
        maxLength={200}
        type="text"
      />
    </label>
  </div>
</div>
`;
